[SPARK-21617][SQL] Store correct metadata in Hive for altered DS table.

vanzin opened the pull request:
This change fixes two issues:
- when loading table metadata from Hive, restore the "provider" field of
  CatalogTable so DS tables can be identified.
- when altering a DS table in the Hive metastore, make sure to not alter
  the table's schema, since the DS table's schema is stored as a table
  property in those cases.

Also added a new unit test for this issue which fails without this change.

Commit:
[SPARK-21617][SQL] Store correct metadata in Hive for altered DS table.

This change fixes two issues:
- when loading table metadata from Hive, restore the "provider" field of
  CatalogTable so DS tables can be identified.
- when altering a DS table in the Hive metastore, make sure to not alter
  the table's schema, since the DS table's schema is stored as a table
  property in those cases.

Also added a new unit test for this issue which fails without this change.

vanzin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -616,15 +616,24 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
             "compatible way. Updating Hive metastore in Spark SQL specific format."
         logWarning(warningMessage, e)
-        client.alterTable(updatedTable.copy(schema = updatedTable.partitionSchema))
+        client.alterTable(updatedTable.copy(schema = tableToStore.partitionSchema))
```
This is the exception handling code I mentioned in the bug report which seems very suspicious. I had half a desire to just remove it, but maybe someone can explain to me why this code makes sense.

viirya replied:
I think this part is directly related to the logic which converts the table metadata to Spark SQL specific format: https://github.com/apache/spark/blob/f18b905f6cace7686ef169fda7de474079d0af23/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L290-L301

SparkQA commented:
**[Test build #80182 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80182/testReport)** for PR 18824 at commit [`aae3abd`](https://github.com/apache/spark/commit/aae3abd673adc7ff939d842e49d566fa722403a3).

vanzin commented:
Hmm, after I made some changes to the test the whole test suite is failing (although running them individually works). I'll work on that but the fix itself, other than the test, should be correct.

SparkQA commented:
**[Test build #80182 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80182/testReport)** for PR 18824 at commit [`aae3abd`](https://github.com/apache/spark/commit/aae3abd673adc7ff939d842e49d566fa722403a3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80182/
Test FAILed.

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -413,7 +414,10 @@ private[hive] class HiveClientImpl(
+      val properties = Option(h.getParameters).map(_.asScala.toMap).getOrElse(Map())
+
+      val provider = properties.get(HiveExternalCatalog.DATASOURCE_PROVIDER)
+        .orElse(Some(DDLUtils.HIVE_PROVIDER))
```
Previously we don't store provider for Hive serde table. Some existing logic to decide if a table retrieved from metastore is a datasource table may be broken due to this change.

viirya replied:
Oh. Nvm. Looks like we access the key `DATASOURCE_PROVIDER` in `table.properties` for that purpose. This should be safe. Anyway, actually we will set `provider` for `CatalogTable` later when restoring the table read from metastore.

Another concern is we previously don't restore `provider` for a view, please refer to https://github.com/apache/spark/blob/f18b905f6cace7686ef169fda7de474079d0af23/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L682. By this change, we will set `provider` to `HIVE_PROVIDER` too for view.

vanzin replied:
> Maybe this is redundant.

This was definitely not redundant in my testing. The metadata loaded from the metastore in `HiveExternalCatalog.alterTableSchema` was definitely not set when I debugged this. In fact the test I wrote fails if I remove this code (or comment the line that sets "provider" a few lines below).

Perhaps some other part of the code sets it in a different code path, but this would make that part of the code redundant, not the other way around.

vanzin replied:
The restoring you mention is done in `HiveExternalCatalog.restoreTableMetadata`. Let me see if I can use that instead of making this change.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -616,15 +616,24 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
+    // If it's a data source table, make sure the original schema is left unchanged; the
+    // actual schema is recorded as a table property.
+    val tableToStore = if (DDLUtils.isDatasourceTable(updatedTable)) {
+      updatedTable.copy(schema = rawTable.schema)
```
We do support `ALTER TABLE ADD COLUMN`, which relies on `alterTableSchema `. The data source tables can be read by Hive if possible.  Thus, I think we should not set the schema unchanged.

gatorsmile replied:
I just checked the JIRA description. This sounds a bug we need to resolve. It needs a little bit complex to fix it. We need to follow what we did for `create table`. cc @xwu0226 Please help @vanzin address this issue.

vanzin replied:
Hmm, I see that this will break DS tables created with `newHiveCompatibleMetastoreTable` instead of `newSparkSQLSpecificMetastoreTable`.

For the former, the only thing I can see that could be used to identify the case is the presence of serde properties in the table metadata. That could replace the `DDLUtils.isDatasourceTable(updatedTable)` check to see whether the schema needs to be updated.

For the latter case, I see that `newSparkSQLSpecificMetastoreTable` stores the partition schema as the table's schema (which sort of explains the weird exception handling I saw). So this code is only correct if the partition schema cannot change. Where is the partition schema for a DS table defined? Is that under control of the user (or the data source implementation)? Because if it can change you can run into pretty much the same issue.

Commit:
Rework the patch.

- Use the same code to translate between Spark and Hive tables when
  creating or altering the table.
- Fix the test so that it doesn't try to create a new SparkSession,
  which conflicts with TestHiveSingleton.
- Use 2.1's EnvironmentContext to disable auto updating of stats
  for DS tables.

vanzin commented:
I reworked the patch to try to merge the "create table" and "alter table" paths, so they both do the translation the same way.

There are still some test failures but I wanted to get this up here for you guys to take a look while I fix those.

SparkQA commented:
**[Test build #80217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80217/testReport)** for PR 18824 at commit [`cc7cd95`](https://github.com/apache/spark/commit/cc7cd955e262e4b45730a6a909129b0c059b15ac).

SparkQA commented:
**[Test build #80217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80217/testReport)** for PR 18824 at commit [`cc7cd95`](https://github.com/apache/spark/commit/cc7cd955e262e4b45730a6a909129b0c059b15ac).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80217/
Test FAILed.

vanzin commented:
FYI, I'm rebuilding the environment where I found the bug, to see why the code was failing even with the exception handler. I'll update the bug if necessary.

vanzin commented:
I updated the bug, let me close this for now while I figure out why that exception is happening.
